site/src/comparison.rs

```diff
@@ -0,0 +1,583 @@
+    loop {
+        let comparison = compare(
+            before,
+            after.clone(),
+            "instructions:u".to_owned(),
+            data,
+            &master_commits,
+        )
+        .await?;
+        log::info!(
+            "Comparing {} to {}",
+            comparison.b.commit,
+            comparison.a.commit
+        );
+
+        // handle results of comparison
+        populate_report(&comparison, &mut report).await;
+
+        // Check that there is a next commit and that the
+        // after commit is not equal to `end`
+        match comparison.next(&master_commits).map(Bound::Commit) {
+            Some(next) if Some(&after) != end.as_ref() => {
+                before = after;
+                after = next;
+            }
+            _ => break,
+        }
+    }
+    let end = end.unwrap_or(after);
+
+    let report = generate_report(&start, &end, report);
+    Ok(api::triage::Response(report))
+}
+
+pub async fn handle_compare(
+    body: api::days::Request,
+    data: &InputData,
+) -> Result<api::days::Response, BoxedError> {
+    let commits = rustc_artifacts::master_commits().await?;
+    let comparison =
+        crate::comparison::compare(body.start, body.end, body.stat, data, &commits).await?;
+
+    let conn = data.conn().await;
+    let prev = comparison.prev(&commits);
+    let next = comparison.next(&commits);
+    let is_contiguous = comparison.is_contiguous(&*conn, &commits).await;
+
+    Ok(api::days::Response {
+        prev,
+        a: comparison.a,
+        b: comparison.b,
+        next,
+        is_contiguous,
+    })
+}
+
+async fn populate_report(comparison: &Comparison, report: &mut HashMap<Direction, Vec<String>>) {
+    if let Some(summary) = summarize_comparison(comparison) {
+        if let Some(direction) = summary.direction() {
+            let entry = report.entry(direction).or_default();
+
+            entry.push(summary.write(comparison).await)
+        }
+    }
+}
+
+fn summarize_comparison<'a>(comparison: &'a Comparison) -> Option<ComparisonSummary<'a>> {
+    let mut benchmarks = comparison.get_benchmarks();
+    // Skip empty commits, sometimes happens if there's a compiler bug or so.
+    if benchmarks.len() == 0 {
+        return None;
+    }
+
+    let cmp = |b1: &BenchmarkComparison, b2: &BenchmarkComparison| {
+        b1.log_change()
+            .partial_cmp(&b2.log_change())
+            .unwrap_or(std::cmp::Ordering::Equal)
+    };
+    let lo = benchmarks
+        .iter()
+        .enumerate()
+        .min_by(|&(_, b1), &(_, b2)| cmp(b1, b2))
+        .filter(|(_, c)| c.is_significant() && !c.is_increase())
+        .map(|(i, _)| i);
+    let lo = lo.map(|lo| benchmarks.remove(lo));
+    let hi = benchmarks
+        .iter()
+        .enumerate()
+        .max_by(|&(_, b1), &(_, b2)| cmp(b1, b2))
+        .filter(|(_, c)| c.is_significant() && c.is_increase())
+        .map(|(i, _)| i);
+    let hi = hi.map(|hi| benchmarks.remove(hi));
+
+    benchmarks.clear();
```

Comment:
Why are we clearing out benchmarks here? It's going to get dropped immediately afterwards anyway and is just a vector as far as I can tell?

Reply:
Ah this is a leftover from a previous change. I will remove